### DISCUSSION TITLE
Release candidate for Python.NET 3.0

### DISCUSSION THREAD

**Discussion by filmor** [pythonnet/pythonnet#1837] [Announcements]
## What's Changed
* Match generic and private methods upon runtime reload by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1637
* Require newest available clr-loader by @filmor in https://github.com/pythonnet/pythonnet/pull/1643
* Fix the PyGILState_STATE type by @filmor in https://github.com/pythonnet/pythonnet/pull/1645
* Fix warning regarding undefined __module__ on GC Offset Base by @filmor in https://github.com/pythonnet/pythonnet/pull/1646
* Removed `ShutdownMode`. Now always behaves like original `Reload` by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1638
* When process is exiting, there's no need to save live .NET objects by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1648
* Moved `Py` class into its own file by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1649
* Added a regression test for calling base method from nested class by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1652
* Fixed accessing partially overriden properties from Python by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1650
* Reworked the way .NET objects are constructed from Python by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1651
* Improved support for generic method overloading by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1657
* Use Delegates to access Py_NoSiteFlag by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1659
* Provide `__int__` instance method on enum types to support `int(Enum.Member)` by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1661
* Cleanup `PyBuffer` a bit by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1662
* Support for byref parameters when overriding .NET methods from Python by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1663
* Performance tests with baseline from Pypi by @filmor in https://github.com/pythonnet/pythonnet/pull/1667
* Move code to subdirectories and rename or split up by @filmor in https://github.com/pythonnet/pythonnet/pull/1665
* Added ARM64 CI action by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1669
* Only run CI on pushes to master and on all pull requests by @filmor in https://github.com/pythonnet/pythonnet/pull/1670
* Temporary fix method binder for out parameters by @eirannejad in https://github.com/pythonnet/pythonnet/pull/1672
* Added todo note to ensure CLA is signed by @eirannejad in https://github.com/pythonnet/pythonnet/pull/1674
* Add tests for exception leaking. by @BadSingleton in https://github.com/pythonnet/pythonnet/pull/1679
* We can drop the import hack as we are now using the newer import mechanics by @filmor in https://github.com/pythonnet/pythonnet/pull/1686
* Update CHANGELOG.md by @nobbi1991 in https://github.com/pythonnet/pythonnet/pull/1690
* Allow dynamic `PyObject` conversion to `IEnumerable` by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1697
* add GitHub URL for PyPi by @andriyor in https://github.com/pythonnet/pythonnet/pull/1708
* Make methods of `PyObject` inherited from its base .NET classes GIL-safe by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1711
* Support for `BigInteger` (C#) <-> `PyInt` by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1710
* On shutdown from Python release all slot holders by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1712
* Clear weakref list when reflected object is destroyed by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1758
* Clear weak reference list when an extension type is destroyed by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1761
* Fix layout of `NativeTypeSpec` on 32 bit platforms by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1765
* Implement IConvertible on PyObject by @filmor in https://github.com/pythonnet/pythonnet/pull/1762
* Ensure that codec tests are run by @filmor in https://github.com/pythonnet/pythonnet/pull/1763
* Fix decimal default parameters by @filmor in https://github.com/pythonnet/pythonnet/pull/1773
* Work around potential Mono bug, that hangs the runtime when new threads start by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1779
* Miscellaneous fixes and cleanups by @filmor in https://github.com/pythonnet/pythonnet/pull/1760
* Fix enum codec by @filmor in https://github.com/pythonnet/pythonnet/pull/1621
* Disallow runtime shutdown when the Python error indicator is set by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1780
* Multiple fixes related to Dictionary.Keys bug by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1786
* Drop Python 3.6 support by @filmor in https://github.com/pythonnet/pythonnet/pull/1795
* Move to modern setuptools with pyproject.toml by @filmor in https://github.com/pythonnet/pythonnet/pull/1793
* `Min`/`MaxSupportedVersion` and `IsSupportedVersion` on `PythonEngine` by @filmor in https://github.com/pythonnet/pythonnet/pull/1799
* `Min`/`MaxSupportedVersion` and `IsSupportedVersion` on `PythonEngine` by @lostmsu in https://github.com/pythonnet/pythonnet/pull/1798
* Add explicit tests for vararg call by @filmor in https://github.com/pythonnet/pythonnet/pull/1805
* Fix demo scripts by @filmor in https://github.com/pythonnet/pythonnet/pull/1802
* fixed ForbidPythonThreadsAttribute being ignored in certain scenarios by @filmor in https://github.com/pythonnet/pythonnet/pull/1815
* Add test for enum int conversion by @filmor in https://github.com/pythonnet/pythonnet/pull/1811
* Implement configuring clr from environment by @filmor in https://github.com/pythonnet/pythonnet/pull/1817
* Take GIL when checking if error occurred on shutdown by @filmor in https://github.com/pythonnet/pythonnet/pull/1836

## New Contributors
* @eirannejad made their first contribution in https://github.com/pythonnet/pythonnet/pull/1672
* @nobbi1991 made their first contribution in https://github.com/pythonnet/pythonnet/pull/1690
* @andriyor made their first contribution in https://github.com/pythonnet/pythonnet/pull/1708

**Full Changelog**: https://github.com/pythonnet/pythonnet/compare/v3.0.0-a2...v3.0.0-rc1

<hr /><em>This discussion was created from the release <a href='https://github.com/pythonnet/pythonnet/releases/tag/v3.0.0-rc1'>Release candidate for Python.NET 3.0</a>.</em>

**Comment by filmor** [comment]
@lostmsu Could you please upload the Nuget packages to nuget.org? My user ([filmor](https://www.nuget.org/profiles/filmor)) either lost access or never had it :)

**Reply by lostmsu** [reply]
I (and you too) can easily release a new version on NuGet with `-preview-xxxxx` suffix using the NuGet Preview GitHub action. Already [launched one](https://github.com/pythonnet/pythonnet/actions/runs/2571487481).

But if we do `-rc1` due to the way versions are handled on NuGet we would need to bump the build number up to 3.0.1 after RC1 commit.

**Reply by lostmsu** [reply]
The NuGet now has [3.0.0-preview2022-06-27](https://www.nuget.org/packages/pythonnet/3.0.0-preview2022-06-27) which matches `-rc1` release on PyPi

**Reply by lostmsu** [reply]
@filmor sent you another invite to NuGet pythonnet organization.

**Reply by filmor** [reply]
I'd suggest we just pause preview packages until 3.0 final is released. Every significant patch/code change on the `release` branch is worth a new rc in my book. As soon as it's finalised, we merge back to `master` and bump the version to 3.1.0-dev.

**Reply by lostmsu** [reply]
@filmor I don't think it makes sense to pause monthly/weekly/daily previews. I'd rather have 3.0.1-rc2, 3.0.2-rc3 etc. E.g. since rc1 is cut we should bump the build up by 1 right now.

**Reply by filmor** [reply]
No, the RCs will be named 3.0.0-rc1..n.

**Reply by lostmsu** [reply]
@filmor is there a reason to not adopt the scheme with 3.0.1-rc2?

See https://github.com/semver/semver/issues/200#issuecomment-43932319

**Comment by m-rossi** [comment]
Packages are also available on [conda-forge](conda-forge.org)
```sh
conda install -c conda-forge/label/pythonnet_rc -c conda-forge pythonnet
```

**Reply by filmor** [reply]
Great, thanks!

**Reply by maxcapodi78** [reply]
@filmor why the 3.0.rc1 has been yanked from pypi? do you plan to release soon 3.0?

**Reply by filmor** [reply]
I yanked it because I accidentally left a debug `raise` in the code (check the release branch). It's fixed already and I will be back home tonight where I can create an rc2.

**Comment by filmor** [comment]
There is a new release candidate available now: https://github.com/pythonnet/pythonnet/releases/tag/v3.0.0-rc2. It's already available from Nuget and Pypi. I encourage you all to test it as much as you can.

If we don't get any reports of blocking bugs or regressions, we will release 3.0 on **2022-07-16**. The time until then will be spent on updating documentation.

A small one up front: If you want to use .NET Core, that has been simplified a bit. You can either do
```python
from pythonnet import load
load("coreclr", runtime_config="location/of/your/runtimeconfig.json")
import clr
```
or you can configure the same thing as environment variables:
```
PYTHONNET_RUNTIME=coreclr
PYTHONNET_CORECLR_RUNTIME_CONFIG=location/of/your/runtimeconfig.json
```

**Reply by maxcapodi78** [reply]
That's great. the new method works also very well. A suggestion for future linux version. I've found a package named dotnetcore2 on pypi.org that works well on linux. I've been able to use this as coreclr in linux (CentOS) and seems to work ok (I had to manually create a runtimeconfig.json because it is not in the dotnetcore2 folder but all dlls and dotnet executable are ok). Maybe you may think in the future to add it as requirement for Linux installation and automatically enable by default the coreclr and dotenetcore2 package and leave user the possibilty to customize it if they want.
This is how I made it working for RC2

```
from pythonnet import load
os.path.dirname(__file__)
runtime = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "dotnetcore2", "bin"))
json_file = os.path.abspath(os.path.join(os.path.dirname(__file__), "misc", "myapp.runtimeconfig.json"))
# rt = get_coreclr(json_file, runtime)
# set_runtime(rt)
load("coreclr", runtime_config=json_file, dotnet_root=runtime)
```

**Comment by filmor** [comment]
I published the third (and hopefully last) release candidate just now: https://github.com/pythonnet/pythonnet/releases/tag/v3.0.0-rc3

This one should fix a few regressions regarding construction of primitive types. The plan is still to release on Saturday. The progress is tracked here: https://github.com/pythonnet/pythonnet/issues/1860

**Comment by filmor** [comment]
I published the fourth (and hopefully last ;)) release candidate just now: https://github.com/pythonnet/pythonnet/releases/tag/v3.0.0-rc4

This one (only) fixes a memory leak found by @m-rossi and fixed by @lostmsu (in PR #1873), see #1872.

**Reply by filmor** [reply]
I'll also need a bit more time for finishing up the remaining tasks (mainly docs and "Minor tweaks" in the tracking issue #1860), so in case there are no further blocking bugs discovered, we will release on **2022-07-20**.

**Reply by filmor** [reply]
I didn't have time to work on this, new ETA is **2022-07-25**.

**Reply by maxcapodi78** [reply]
@filmor do you have any update?

**Reply by filmor** [reply]
It's a bit stuck. I have been working on some improvements to clr-loader and the docs are still not in. I'll try to get some work done until the weekend.

**Reply by emilaz** [reply]
Hey @filmor, thanks for putting so much work into this :) 
I was wondering if you have any updates for us regarding the new release date?

**Reply by maxcapodi78** [reply]
Hey @filmor any update about the final ETA?

**Reply by filmor** [reply]
The only thing missing is finishing up the documentation. I hope to be able to do that by next week. Please test rc6 thoroughly.

**Comment by heartacker** [comment]
after update to RC6

we could not use `get_coreclr()` any more. but it works in rc 4

```
TypeError
get_coreclr() takes 0 positional arguments but 1 was given
  File "D:\lourney\__init__.py", line 29, in <module>
    rt = get_coreclr(os.path.join(dllpath, configfile))
```

**Reply by filmor** [reply]
The function parameters are all strict keyword arguments now, so this would need to be `get_coreclr(runtime_config=...)` now.
